docs: revise RELEASE_PROCESS.md (#1086)
open-component-model/ocm: RELEASE_PROCESS.md

In any case, a pre-release may be created by specifying a pre-release suffix for
the release action execution. This will lead (for most use cases) to the
creation of a "Release Candidate" which can be tested
and delivered to end users willing to test the new release.

## Release Workflow

### The Release Branch Creation / Cutoff

Every Minor release starts with the creation of a release branch through [
`Release Branch Creation`](./.github/workflows/release-branch.yaml).

The version / minor of the release is based on the content of the file [
`VERSION`](./VERSION) which
is updated automatically said `release` action. During development, the
content of this file is the complete release name of the release currently under
development and the suffix `-dev` (e.g. `0.1.0-dev`). The content of this file
is used for generating the version information compiled into the ocm
executables.

The release branch is then created with the following steps:

1. `main` is checked out and the [`VERSION`](./VERSION) file is read.
2. The combination of <major>.<minor> is read from VERSION and used to create
   the branch name, e.g. `release/0.17`.
3. The branch is created and pushed to the repository.

> 4. A Pull Request is created by a bot to bump the [`VERSION`](./VERSION) file to

Reviewer: ```suggestion
4. A Pull Request is created by a bot to bump the [`VERSION`](./VERSION) file on `main` to
```

   the next minor version, e.g. `0.18.0-dev`.

At this point in time we call the minor release `0.17` cut-off.

This means that:

- We no longer accept features for the development of this branch
- We no longer accept breaking changes for the development of this branch
- Any change that is not a bug fix or a documentation change must be approved by
  the release manager
- Any bug fix that is not deemed critical must be approved by the release

>   manager

Author: To be defined. We do not have one right now but it is defacto currently the person with the most experience in releasing. I will bring it up in our regular sync

>   manager

Reviewer 2: > I will bring it up in our regular sync

Which sync?

>   manager

Author: Since we dont have community meetings atm, it would be internal only

>   manager

Author: Decision: 
We choose Release Managers (for now internally) on demand for every release based on a random selection process. I will continously pair up with them for the first releases to ensure proper knowledge transfer and fill gaps where needed

- Any bug fix must first be merged to main and then [
  `cherry-picked`](https://git-scm.com/docs/git-cherry-pick) to the release
  branch.

At this point in time, any release targeted on this branch will have this minor
version as a base.

### Preparing a Minor release candidate

After the cut-off, the release manager will usually prepare a release candidate.
This is done by creating a pre-release on the release branch, that goes along
with a qualifying suffix.

Currently we only use one form of suffixed, pre-release, the
`Release Candidate`: Any Release Candidate
is testable by users and signalled in the form of
`<major>.<minor>.<patch>-rc.<rc-number>`.

If a release candidate is created, the `-dev`-suffix is removed and the suffix
is appended to generate the name of the release.

During the release, just before creating the git tag for a release,
the [VERSION](./VERSION) file is changed to include this new suffix.
The transformation thus looks like

> ```shell

Author: Unfortunately it is, otherwise I get a markdownlint issue that the codeblock is unspecified. Can you give me a better idea?

> ```shell

Reviewer: ```suggestion
```text
```

<major>.<minor>.<patch>-dev -> <major>.<minor>.<patch>-rc.<rc-number>
```

### Creating a Minor release

> Once a release candidate is seen as sufficiently tested, the release manager can

Author: I am open for adding a TODO if you tell me what we want here in the future

to stakeholders),
and after a grace period, promote the draft release to a full release.

This promotion is currently effectively a full rebuild from the release branch,
with the difference that the `-rc.<rc-number>` suffix is removed.

After the build, instead of finishing, the [
`release`](./.github/workflows/release.yaml) GitHub Action will also publish the
release.

This publishing to package registries (such as brew) is delegated to [
`publish-to-other-than-github`](./.github/workflows/publish-to-other-than-github.yaml)

After the release on a release branch was successful, because the release was
now run without a suffix, the version is considered `burned`.
This means that, even if bugs are found for that release in the future, a
`patch` release will be created for that release branch.

Concretely this means that the following additional steps are taken:

1. The release is tagged with the version number from the [`VERSION`](./VERSION)
   file, *without* the `-rc.<rc-number>` suffix.
2. The release is published on GitHub as the latest release, not as a
   pre-release.
3. The release is published to the package registries.

> 4. The release branch is bumped in [`VERSION`](./VERSION) to the next <patch>,

Reviewer: ```suggestion
4. The [`VERSION`](./VERSION) is bumped in the release branch to the next <patch>,
```

   e.g. `0.17.0-dev` becomes `0.17.1-dev` via Pull Request.

### Creating a Patch release

> The process to creating a Patch release is almost equivalent to the process of

Reviewer: ```suggestion
The process to create a Patch release is almost equivalent to the process of
```

creating a Minor release.

> Whenever a patch release in the form of <major>.<minor>.<patch> is needed, the

Author: ```suggestion
Whenever a patch release in the form of `<major>.<minor>.<patch>` is needed, the
```
?

> Whenever a patch release in the form of <major>.<minor>.<patch> is needed, the

Reviewer 2: this should work

branch `releases/<major>.<minor>` is used to prepare the release.
The only difference is that now the VERSION file should contain the suffix
`<major>.<minor>.<patch>-dev`
(which should have automatically been bumped on the last release).

This means that creating candidates for the patch or creating the release is
equivalent by using the [`release`](./.github/workflows/release.yaml) GitHub

> Action.

Reviewer: https://github.com/open-component-model/ocm/pull/1090#discussion_r1843534797

Author: How does this comment relate to this one here?

Reviewer: the release action did bump version wrongly... now it should work I guess

Make sure that all patched commits have been cherry-picked from main.

*NOTE: It is not valid to create a fresh commit on a patch branch without a

> corresponding cherry-pick from main. Use the next minor version for features.*

Reviewer: > Use the next minor version for features.

I don't get that... context is patch release process... which doesn't allow features...

Author: youre right its obvious, i removed that part

> #### How to cherry-pick a commit from Main to a Patch branch

Author: thats true, i changed it to release branch

To cherry-pick a commit from `main` to a patch branch, the following steps are
necessary:

1. Checkout the release branch for the patch, e.g. `releases/0.17` for a patch
   release of `0.17.0` towards `0.17.1`.
2. Cherry-pick the commit from `main` to the patch branch, e.g.
   `git cherry-pick <commit-hash>`. Resolve conflicts as necessary.
3. Create a Pull-Request for the cherry-picked commit to the branch,

>    and prefix it with `[releases/<major>.<minor>]` (in this case

Reviewer: ```suggestion
   and prefix the title with `[releases/<major>.<minor>]` (in this case
```

   `[releases/0.17]`)
   to signal that this is a patch commit, e.g. with `gh`

   ```shell
   gh pr create \
      --title "[releases/0.17] cherry-pick: <Original PR or Commit>" \
      --body "Cherry-pick of <Original PR or Commit> from main to releases/0.17" \
      --base releases/0.17 \
      --draft
   ```

4. Merge the Pull-Request to the patch branch.

### How Release Notes are managed

For every release branch there is a Github Action
called [Release Drafter](./.github/workflows/release-drafter.yaml).
This workflow interprets the Pull Requests merged against main and the release
branches and generates a draft release
in Github which can be formed and edited.

Note that when you are updating the branches, the release notes currently get
overwritten from scratch so any edits get lost.
In case you want to permanently change the release notes, you will have to carry
them through all release candidates manually.

> (TODO: this needs improvement by allowing us to do "append-only" style release

Author: yes it should I created https://github.com/open-component-model/ocm/issues/1097 as a follow up and linked it

notes)

When a release is created, the release notes are copied to
the [./docs/releasenotes](./docs/releasenotes) folder in that branch
with a dedicated commit. If the release notes for that branch already exist in
the folder, the release notes are taken from the file
instead of the draft notes from GitHub. This means that you can also create a
Pull Request to the branch with the corresponding release notes
and it will be used for the release.
(TODO: Currently we have not decided if we want to permanently keep the notes
and thus need a PR to main to append them, or if we want to drop them in the

> future.)

Reviewer: @morri-son - didn't we agree already on throwing them away?

Author: We originally did, but due to the issues we have with release note maintenance we will have to tackle this in a separate issue that likely requires rewriting how we handle the release note generation. I created https://github.com/open-component-model/ocm/issues/1097 to follow up on this specifically because we might even need to change how we generate the notes

uploaded to the respective package repositories. All currently supported
installation methods are described
[here](https://github.com/open-component-model/ocm?tab=readme-ov-file#installation).

### Addendum

This release process got rewritten as of 0.19.0 and thus earlier releases
followed model where release branches
were kept in the form of `releases/<major>.<minor>.<patch>` and the release
branches were created on demand with the need
of branches. This model has now largely been replaced by the model we see in

> this document

Reviewer: ```suggestion
this document.
```
